# Table of content

1. Importing libraries
2. Importing data
3. Data wrangling
4. Subsetting
5. Data Analysis
6. Exporting datasets

# 1. Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing data

In [3]:
# Define path
path = r'/Users/frederikeschulz-mullensiefen/Desktop/Master Folder_Instacart/02_Data'

In [4]:
# Define columns orders data
vars_list = ['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day', 'days_since_prior_order']

In [5]:
# Import orders data
df_orders = pd.read_csv(os.path.join(path, 'Original Data', 'orders.csv'), usecols = vars_list)

In [6]:
# Import products data
df_products = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

In [7]:
# Import department data
df_departments = pd.read_csv(os.path.join(path, 'Original Data', 'departments.csv'), index_col = False)

# 3. Data wrangling

3.1 Checking for missing values orders data

In [8]:
# Checking for missing values 
df_orders['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

3.2 Renaming Columns

In [9]:
# Renaming order_dow into orders_day_of_week
df_orders.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [10]:
# Renaming order_hour_of_day into ordertime_hour_of_day
df_orders.rename(columns = {'order_hour_of_day' : 'ordertime_hour_of_day'}, inplace = True)

In [11]:
df_orders

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         ordertime_hour_of_day  days_since_prior_order  
0                            8                     NaN  
1                            7                    15.0  
2                           12                    21.0  
3                            7                    29.0  
4                           15                    28.0  
...                        ...                     ...  
3421078                     18                    29.0  
3421079                     10                    30.0  
3421080                     12                    18.0  
3421081                     12                     7.0  
3421082                     14                    30.0  

[3421083 rows x 6 columns]

3.3 Changing data types

In [12]:
# Changing order_id into string data type
df_orders['order_id'] = df_orders['order_id'].astype('str')

In [13]:
# Changing user_id into string data type
df_orders['user_id'] = df_orders['user_id'].astype('str')

3.4 Transposing department data

In [14]:
# Transposing department data
df_departments.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [15]:
# Creating new version of transposed dataframe
df_departments_t = df_departments.T

In [16]:
# Changing index
df_departments_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [17]:
# Defining new header
new_header = df_departments_t.iloc[0]

In [18]:
# Defining data 
df_departments_t_new = df_departments_t[1:]

In [19]:
# Inserting new header
df_departments_t_new.columns = new_header

3.3 Creating a data dictionary

In [20]:
# Creating data dictionary
data_dict = df_departments_t_new.to_dict('index')

In [21]:
df_products[df_products['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# 4. Subsetting

In [22]:
# Creating new dataframe for snacks
df_snacks = df_products[df_products['department_id']==19]

In [23]:
# Creating new dataframe for breakfast
df_breakfast = df_products[df_products['department_id']==14]

In [24]:
# Creating new dataframe for dinner parties
df_dinnerparties = df_products.loc[df_products['department_id'].isin([5,20,7,12])]

# 5. Data Analysis

In [25]:
# Analysis busiest hour for placing orders
df_orders['ordertime_hour_of_day'].mode()

0    10
Name: ordertime_hour_of_day, dtype: int64

The busiest hour for placing orders is 10:00h.

In [26]:
# Meaning department_id 4
print(data_dict.get('4'))

{'department': 'produce'}


The meaning behind the department_id 4 is produce

In [27]:
# Rows of dinner parties dataframe
df_dinnerparties.shape

(7650, 5)

The last dataframe created (df_dinnerparties) has 7650 rows. 

In [28]:
# Information about customer with user_id = 1
df_orders[df_orders['user_id']== '1']

order_id user_id  order_number  orders_day_of_week  ordertime_hour_of_day  \
0   2539329       1             1                   2                      8   
1   2398795       1             2                   3                      7   
2    473747       1             3                   3                     12   
3   2254736       1             4                   4                      7   
4    431534       1             5                   4                     15   
5   3367565       1             6                   2                      7   
6    550135       1             7                   1                      9   
7   3108588       1             8                   1                     14   
8   2295261       1             9                   1                     16   
9   2550362       1            10                   4                      8   
10  1187899       1            11                   4                      8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [29]:
# Getting statistics for customer with user_id = 1
x = df_orders[df_orders['user_id']== '1']

In [30]:
# Count, Mean, Standard Deviation , Min, Max, Quartiles of user_id = 1
x.describe()

order_number  orders_day_of_week  ordertime_hour_of_day  \
count     11.000000           11.000000              11.000000   
mean       6.000000            2.636364              10.090909   
std        3.316625            1.286291               3.477198   
min        1.000000            1.000000               7.000000   
25%        3.500000            1.500000               7.500000   
50%        6.000000            3.000000               8.000000   
75%        8.500000            4.000000              13.000000   
max       11.000000            4.000000              16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [31]:
# Mode of ordertime of user_id = 1
x['ordertime_hour_of_day'].mode()

0    7
1    8
Name: ordertime_hour_of_day, dtype: int64

In [32]:
# Mode of order day of user_id = 1
x['orders_day_of_week'].mode()

0    4
Name: orders_day_of_week, dtype: int64

# 6. Exporting datasets

In [36]:
# Exporting orders dataframe 
df_orders.to_csv(os.path.join(path,'Prepared Data', 'orders_wrangled.csv'), index = False)

In [34]:
# Exporting departments dataframe
df_departments_t_new.to_csv(os.path.join(path,'Prepared Data', 'departments_wrangled.csv'))

In [35]:
df_orders

order_id user_id  order_number  orders_day_of_week  \
0        2539329       1             1                   2   
1        2398795       1             2                   3   
2         473747       1             3                   3   
3        2254736       1             4                   4   
4         431534       1             5                   4   
...          ...     ...           ...                 ...   
3421078  2266710  206209            10                   5   
3421079  1854736  206209            11                   4   
3421080   626363  206209            12                   1   
3421081  2977660  206209            13                   1   
3421082   272231  206209            14                   6   

         ordertime_hour_of_day  days_since_prior_order  
0                            8                     NaN  
1                            7                    15.0  
2                           12                    21.0  
3                            7                    29.0  
4                           15                    28.0  
...                        ...                     ...  
3421078                     18                    29.0  
3421079                     10                    30.0  
3421080                     12                    18.0  
3421081                     12                     7.0  
3421082                     14                    30.0  

[3421083 rows x 6 columns]